crates/starknet-types-core/src/felt/secret_felt.rs

```diff
@@ -0,0 +1,280 @@
+    ///
+    /// let mut private_key = Felt::from_hex_unchecked("0x123...");
+    /// let secret_felt = SecretFelt::from_felt(&mut private_key);
+    /// // private_key is now zeroized (set to Felt::ZERO)
+    /// ```
+    pub fn from_felt(secret_felt: &mut Felt) -> Self {
+        let boxed_copy = Box::new(*secret_felt);
+        secret_felt.zeroize();
+        Self(boxed_copy)
+    }
+
+    /// Creates a new [SecretFelt] from a hex String and zeroized the original String.
+    ///
+    /// # Example
+    /// ```
+    /// use std::fs;
+    /// use starknet_types_core::felt::secret_felt::SecretFelt;
+    /// use std::str::FromStr;
+    ///
+    /// // make sure the String is initialized in a secure way
+    /// let mut private_key = fs::read_to_string("path/to/secret_value").unwrap();
+    /// let secret_felt = SecretFelt::from_hex_string(&mut private_key);
+    /// ```
+    pub fn from_hex_string(hex: &mut String) -> Result<Self, FromStrError> {
+        let secret_felt = Felt::from_hex(hex)?;
+        hex.zeroize();
+        Ok(Self(Box::new(secret_felt)))
+    }
+
+    /// Creates a new [SecretFelt] from its big-endian representation in a Vec<u8> of length 32.
+    /// Internally it uses [from_bytes_be](Felt::from_bytes_be).
+    /// The input will be zeroized after calling this function
+    pub fn from_bytes_be(secret: &mut [u8; 32]) -> Self {
+        let secret_felt = Self(Box::new(Felt::from_bytes_be(secret)));
+        secret.zeroize();
+        secret_felt
+    }
+
+    /// Creates a new [SecretFelt] from its little-endian representation in a Vec<u8> of length 32.
+    /// Internally it uses [from_bytes_le](Felt::from_bytes_le).
+    /// The input will be zeroized after calling this function
+    pub fn from_bytes_le(secret: &mut [u8; 32]) -> Self {
+        let secret_felt = Self(Box::new(Felt::from_bytes_le(secret)));
+        secret.zeroize();
+        secret_felt
+    }
+
+    /// Create a new [SecretFelt] from cryptographically secure PRNG
+    ///
+    /// # Example
+    /// ```
+    /// use starknet_types_core::felt::secret_felt::SecretFelt;
+    /// use rand_chacha::ChaCha20Rng;
+    /// use rand::SeedableRng;
+    ///
+    /// let rng = ChaCha20Rng::from_os_rng();
+    /// let secret_key = SecretFelt::from_random(rng);
+    /// ```
+    pub fn from_random<T>(mut rng: T) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        let mut buffer = [0u8; 32];
+        rng.fill_bytes(&mut buffer);
+
+        let secret_felt = Self(Box::new(Felt::from_bytes_be(&buffer)));
+        buffer.zeroize();
+
+        secret_felt
+    }
+
+    /// Returns a safe copy of the inner value.
+    ///
+    /// # Warning
+    ///
+    /// Be careful not to copy the value elsewhere, as that would defeat
+    /// the security guarantees of this type.
+    pub fn inner_value(&self) -> Zeroizing<Felt> {
+        Zeroizing::new(*self.0.clone())
+    }
+}
+
+/// Constant time equality check for [SecretFelt]
+impl PartialEq for SecretFelt {
```

Comment:
I think we should also implement `Eq`.
The additional reflexivity requirement is fulfilled for `Felt`

https://doc.rust-lang.org/std/cmp/trait.Eq.html

Am I missing something?